core/src/main/java/org/apache/iceberg/data/avro/PlannedDataReader.java

```diff
@@ -97,7 +98,8 @@ public ValueReader<?> record(Type partner, Schema record, List<ValueReader<?>> f
 
       Types.StructType expected = partner.asStructType();
       List<Pair<Integer, ValueReader<?>>> readPlan =
-          ValueReaders.buildReadPlan(expected, record, fieldReaders, idToConstant);
+          ValueReaders.buildReadPlan(
+              expected, record, fieldReaders, idToConstant, GenericDataUtil::internalToGeneric);
```

Comment:
This change is only needed for the _generic_ data model and not for the _internal_ data model because defaults are already using the correct classes for internal.